service/controller/network/v1/resource/ipam/create.go

```diff
@@ -1,9 +1,62 @@
 package ipam
 
-import "context"
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/meta"
+
+	"github.com/giantswarm/cluster-operator/service/controller/network/v1/key"
+)
 
 // EnsureCreated takes care of cluster subnet allocation when
 // ClusterNetworkConfig object is created.
 func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	clusterNetworkCfg, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	clusterID := key.ClusterID(clusterNetworkCfg)
+
+	if clusterNetworkCfg.Status.IP != "" {
+		// Subnet allocated. No need to do anything.
+		r.logger.LogCtx(ctx, "level", "debug", "message", "Subnet allocated. No need to do anything.", "clusterID", clusterID)
+		return nil
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "Subnet not allocated. Allocating.", "clusterID", clusterID)
+
+	maskBits := key.ClusterNetworkMaskBits(clusterNetworkCfg)
+
+	mask := net.CIDRMask(maskBits, 32)
+
+	subnet, err := r.ipam.CreateSubnet(ctx, mask, clusterID)
+	if err != nil {
+		return microerror.Maskf(err, "Subnet allocation failed with mask %s. clusterID: %s", net.IP(mask).String(), clusterID)
+	}
+
+	clusterNetworkCfg.Status.IP = subnet.IP.String()
+	clusterNetworkCfg.Status.Mask = net.IP(subnet.Mask).String()
+
+	_, err = r.g8sClient.CoreV1alpha1().ClusterNetworkConfigs(accessor.GetNamespace()).UpdateStatus(&clusterNetworkCfg)
+	if err != nil {
+		ipamDeleteErr := r.ipam.DeleteSubnet(ctx, subnet)
```

Comment:
That smells for me. The operator should try to reconcile towards the desired state until it succeeds. Adding and removing state back and forth feels like a bigger design issue which is here probably related to the storage usage which I really dislike. The network range allocation should be idempotent as I argued already a couple of times and should be bound to a specific namespace or identifier. This is the inherent flaw of the IPAM implementation as we had it so far and now it looks like we continue using it but just let the operator execute the code. This is not an improvement to me. We should rely on the actual state of the system we work against with. Which in this case would be the provider. Either Azure or AWS.

Reply:
Regarding this, I already commented earlier.